datadog_checks_base/datadog_checks/base/utils/http.py

```diff
@@ -322,14 +331,27 @@ def _request(self, method, url, options):
             persist = True  # UDS support is only enabled on the shared session.
             url = quote_uds_url(url)
 
+        self.handle_auth_token(method=method, url=url, default_options=self.options)
+
         with ExitStack() as stack:
             for hook in self.request_hooks:
                 stack.enter_context(hook())
 
-            if persist:
-                return getattr(self.session, method)(url, **new_options)
+            try:
+                if persist:
+                    response = getattr(self.session, method)(url, **new_options)
+                else:
+                    response = getattr(requests, method)(url, **new_options)
+            except Exception as e:
+                self.handle_auth_token(method=method, url=url, default_options=self.options, error=str(e))
+                raise e
             else:
-                return getattr(requests, method)(url, **new_options)
+                try:
+                    response.raise_for_status()
+                except Exception as e:
+                    self.handle_auth_token(method=method, url=url, default_options=self.options, error=str(e))
```

Comment:
Why do we call/retry `self.handle_auth_token` again here ? It's already called just before at line 334 🤔

Reply:
We do it on errors too, which usually means the token needs refreshing

Reply:
1/ Could you add a comment this is needed for token refresh ?

2/ About:

```
                try:
                    response.raise_for_status()
                except Exception as e:
                    self.handle_auth_token(method=method, url=url, default_options=self.options, error=str(e))
```

We don't retry if there is an error ?


3/ I think it would be cleaner to separate case where there is an `self.auth_token_handler` and case where there is no `self.auth_token_handler`. Example:

```
            if persist:
                method_fn = getattr(self.session, method)
            else:
                method_fn = getattr(requests, method)

            if self.auth_token_handler:
                try:
                    response = method_fn(url, **new_options)
                    response.raise_for_status()
                except Exception as e:
                    self.handle_auth_token(method=method, url=url, default_options=self.options, error=str(e))
                    response = method_fn(url, **new_options)
            else:
                response = method_fn(url, **new_options)
            return response
```
```

Reply:
For 3/, wouldn't the first run always fail?